File: Mobile/lib/modules/field_activities/models/field_activity.dart
class FieldActivity {
  const FieldActivity({
    required this.id,
    required this.activityType,
    required this.activityName,
    this.employeeId,
    this.employeeName,
    this.staffRoleLabel,
    this.centerId,
    this.centerName,
    this.activityTypeLabel,
    this.activityAt,
    this.activityDate,
    this.activityTime,
    this.remarks,
    this.photoUrl,
    this.latitude,
    this.longitude,
    this.location,
    this.mapsUrl,
  });

  final int id;
  final int? employeeId;
  final String? employeeName;
  final String? staffRoleLabel;
  final int? centerId;
  final String? centerName;
  final String activityType;
  final String? activityTypeLabel;
  final String activityName;
  final DateTime? activityAt;
  final String? activityDate;
  final String? activityTime;
  final String? remarks;
  final String? photoUrl;
  final double? latitude;
  final double? longitude;
  final String? location;
  final String? mapsUrl;

  factory FieldActivity.fromJson(Map<String, dynamic> json) {
    return FieldActivity(
      id: int.tryParse('${json['id']}') ?? 0,
      employeeId: int.tryParse('${json['employee_id'] ?? ''}'),
      employeeName: json['employee_name']?.toString(),
      staffRoleLabel: json['staff_role_label']?.toString(),
      centerId: int.tryParse('${json['center_id'] ?? ''}'),
      centerName: json['center_name']?.toString(),
      activityType: json['activity_type']?.toString() ?? 'other',
      activityTypeLabel: json['activity_type_label']?.toString(),
      activityName:
          json['activity_name']?.toString() ??
          json['activity_type_label']?.toString() ??
          'Activity',
      activityAt: DateTime.tryParse('${json['activity_at'] ?? ''}'),
      activityDate: json['activity_date']?.toString(),
      activityTime: json['activity_time']?.toString(),
      remarks: json['remarks']?.toString(),
      photoUrl: json['photo_url']?.toString(),
      latitude: double.tryParse('${json['latitude'] ?? ''}'),
      longitude: double.tryParse('${json['longitude'] ?? ''}'),
      location: json['location']?.toString(),
      mapsUrl: json['maps_url']?.toString(),
    );
  }

  String get displayName =>
      activityName.trim().isEmpty ? (activityTypeLabel ?? 'Activity') : activityName;

  String get whenLabel {
    final date = activityDate;
    final time = activityTime;
    if ((date ?? '').isNotEmpty && (time ?? '').isNotEmpty) {
      return '$date · $time';
    }
    if ((date ?? '').isNotEmpty) return date!;
    if (activityAt == null) return '—';
    return activityAt!.toIso8601String();
  }
}

class FieldActivityEmployeeOption {
  const FieldActivityEmployeeOption({required this.id, required this.name});

  final int id;
  final String name;

  factory FieldActivityEmployeeOption.fromJson(Map<String, dynamic> json) {
    return FieldActivityEmployeeOption(
      id: int.tryParse('${json['id']}') ?? 0,
      name: json['full_name']?.toString() ?? 'Employee',
    );
  }
}

class FieldActivityListResult {
  const FieldActivityListResult({
    required this.items,
    this.employees = const [],
  });

  final List<FieldActivity> items;
  final List<FieldActivityEmployeeOption> employees;
}

File: Backend/app/Models/FieldActivity.php
<?php

namespace App\Models;

use App\Enums\FieldActivityType;
use App\Support\AttendanceCalendar;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class FieldActivity extends Model
{
    protected $fillable = [
        'employee_id',
        'user_id',
        'center_id',
        'scheme_id',
        'activity_type',
        'activity_name',
        'activity_at',
        'remarks',
        'photo_path',
        'latitude',
        'longitude',
        'location',
    ];

    protected function casts(): array
    {
        return [
            'activity_type' => FieldActivityType::class,
            'activity_at' => 'datetime',
            'latitude' => 'decimal:7',
            'longitude' => 'decimal:7',
        ];
    }

    public function employee(): BelongsTo
    {
        return $this->belongsTo(Employee::class);
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function center(): BelongsTo
    {
        return $this->belongsTo(Center::class);
    }

    public function scheme(): BelongsTo
    {
        return $this->belongsTo(Scheme::class);
    }

    public function mapsUrl(): ?string
    {
        return Attendance::googleMapsUrl($this->latitude, $this->longitude);
    }

    public function photoUrl(): ?string
    {
        return \App\Support\PublicStorage::url($this->photo_path);
    }

    /**
     * @return array<string, mixed>
     */
    public function toApiArray(): array
    {
        $type = $this->activity_type instanceof FieldActivityType
            ? $this->activity_type
            : FieldActivityType::tryFromMixed((string) $this->activity_type);
        $at = $this->activity_at?->timezone(AttendanceCalendar::TIMEZONE);

        return [
            'id' => $this->id,
            'employee_id' => $this->employee_id,
            'employee_name' => $this->employee?->full_name,
            'staff_role' => $this->employee?->staff_role,
            'staff_role_label' => $this->employee?->staffRoleEnum()->label(),
            'center_id' => $this->center_id,
            'center_name' => $this->center?->name,
            'scheme_id' => $this->scheme_id,
            'scheme_name' => $this->scheme?->name,
            'activity_type' => $type->value,
            'activity_type_label' => $type->label(),
            'activity_name' => $this->activity_name ?: $type->label(),
            'activity_at' => $at?->toIso8601String(),
            'activity_date' => $at?->toDateString(),
            'activity_time' => $at?->format('h:i A'),
            'remarks' => $this->remarks,
            'photo_path' => $this->photo_path,
            'photo_url' => $this->photoUrl(),
            'latitude' => $this->latitude,
            'longitude' => $this->longitude,
            'location' => $this->location,
            'maps_url' => $this->mapsUrl(),
        ];
    }
}
